### ISSUE TITLE
Update Python to 3.9.x

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#4170]
First commit, upgrade Python.

~In second commit, combine usages of `@classmethod` with `@property` where possible, since support for that has been [added](https://docs.python.org/3.9/library/functions.html#classmethod) in 3.9.~ Blocked by https://youtrack.jetbrains.com/issue/PY-47615, postponed as https://github.com/DataBiosphere/azul/issues/4182

In third commit, change `azul.cache` to use `functools.cache` instead of `.lru_cache` since the former is [more efficient](https://docs.python.org/3/library/functools.html#functools.cache).

Can't go all the way to 3.10 due to lack of support in AWS Lambda (https://github.com/aws/aws-lambda-base-images/issues/31).
